be/src/exprs/agg/array_agg.h

```diff
@@ -2,39 +2,88 @@
 
 #pragma once
 
+#include <set>
+
 #include "column/array_column.h"
 #include "column/column_helper.h"
+#include "column/hash_set.h"
 #include "column/type_traits.h"
 #include "exprs/agg/aggregate.h"
 #include "runtime/mem_pool.h"
 #include "runtime/primitive_type.h"
+#include "udf/udf.h"
+#include "udf/udf_internal.h"
 
 namespace starrocks::vectorized {
 
-template <PrimitiveType PT>
+template <PrimitiveType PT, bool is_distinct, typename MyHashSet = std::set<int>>
 struct ArrayAggAggregateState {
     using ColumnType = RunTimeColumnType<PT>;
+    using CppType = RunTimeCppType<PT>;
+    using KeyType = typename SliceHashSet::key_type;
+    void update(MemPool* mem_pool, const ColumnType& column, size_t offset, size_t count) {
+        if constexpr (is_distinct) {
+            if constexpr (pt_is_string<PT>) {
+                for (int i = 0; i < count; i++) {
+                    auto raw_key = column.get_slice(offset + i);
+                    if (set.find(raw_key) == set.end()) {
+                        data_column.append(column, offset + i, 1);
+                        KeyType key(raw_key);
+                        set.template lazy_emplace(key, [&](const auto& ctor) {
```

Comment:
why not directly call lazy_emplace? this way, you call find_or_prepare_insert method twice.

Reply:
refactor it by just emplace at update, and convert set to data column when serialize or finalize.